File: instance/src/lib/AdminClient.js
'use strict';

const { request } = require('undici');

class AdminClient {
  constructor({ baseUrl, adminToken, idInstance, logger, timeoutMs = 10000 }) {
    if (!baseUrl) throw new Error('AdminClient: baseUrl required');
    if (!adminToken) throw new Error('AdminClient: adminToken required');
    if (!idInstance) throw new Error('AdminClient: idInstance required');
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.adminToken = adminToken;
    this.idInstance = idInstance;
    this.logger = logger || console;
    this.timeoutMs = timeoutMs;
  }

  async _req(method, path, body) {
    const url = `${this.baseUrl}/api${path}`;
    const headers = {
      'X-Admin-Token': this.adminToken,
      'Content-Type': 'application/json',
    };

    for (let attempt = 0; attempt < 2; attempt++) {
      try {
        const res = await request(url, {
          method,
          headers,
          body: body ? JSON.stringify(body) : undefined,
          bodyTimeout: this.timeoutMs,
          headersTimeout: this.timeoutMs,
        });

        const text = await res.body.text();
        const data = text ? JSON.parse(text) : null;

        if (res.statusCode >= 500 && attempt === 0) {
          this.logger.warn({ url, status: res.statusCode }, 'AdminClient: 5xx, retry');
          await new Promise((r) => setTimeout(r, 1000));
          continue;
        }

        if (res.statusCode >= 400) {
          throw new Error(`AdminClient ${method} ${path}: ${res.statusCode} ${text}`);
        }
        return data;
      } catch (err) {
        if (attempt === 0) {
          this.logger.warn({ err: err.message }, 'AdminClient: transport error, retry');
          await new Promise((r) => setTimeout(r, 1000));
          continue;
        }
        throw err;
      }
    }
  }

  register({ pid, version, ipv6, state }) {
    return this._req('POST', `/instances/${this.idInstance}/register`, { pid, version, ipv6, state });
  }

  heartbeat({ state, lastEventAt }) {
    return this._req('POST', `/instances/${this.idInstance}/heartbeat`, { state, lastEventAt });
  }

  sendQr({ qr, expiresAt, kind = 'qr' }) {
    return this._req('POST', `/instances/${this.idInstance}/qr`, { qr, expiresAt, kind });
  }

  getConfig() {
    return this._req('GET', `/instances/${this.idInstance}/config`);
  }

  stateChange({ from, to, reason }) {
    return this._req('POST', `/instances/${this.idInstance}/state-change`, { from, to, reason });
  }
}

module.exports = AdminClient;

File: instance-telegram/src/lib/admin_client.py
import logging
from typing import Any, Optional

import httpx


class AdminClient:
    def __init__(self, base_url: str, admin_token: str, id_instance: str,
                 logger: Optional[logging.Logger] = None) -> None:
        self.base_url = base_url.rstrip("/")
        self.token = admin_token
        self.id_instance = id_instance
        self.log = logger or logging.getLogger("admin-client")

    async def _req(self, method: str, path: str, json: Optional[dict] = None) -> Optional[dict]:
        url = f"{self.base_url}{path}"
        headers = {"Authorization": f"Bearer {self.token}"}
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                resp = await client.request(method, url, headers=headers, json=json)
                if resp.status_code >= 400:
                    self.log.warning("admin %s %s -> %d %s", method, path, resp.status_code, resp.text[:200])
                    return None
                if resp.headers.get("content-type", "").startswith("application/json"):
                    return resp.json()
                return {"ok": True}
        except Exception as e:
            self.log.warning("admin %s %s failed: %s", method, path, e)
            return None

    async def get_config(self) -> Optional[dict]:
        return await self._req("GET", f"/instances/{self.id_instance}/config")

    async def send_qr(self, qr: str, expires_at: int, kind: str) -> None:
        await self._req("POST", f"/instances/{self.id_instance}/qr",
                        {"qr": qr, "expiresAt": expires_at, "kind": kind})

    async def heartbeat(self, state: str, wid: Optional[str] = None,
                        phone_number: Optional[str] = None) -> None:
        payload: dict[str, Any] = {"state": state}
        if wid is not None:
            payload["wid"] = wid
        if phone_number is not None:
            payload["phoneNumber"] = phone_number
        await self._req("POST", f"/instances/{self.id_instance}/heartbeat", payload)

    async def state_change(self, from_: str, to: str, reason: Optional[str] = None) -> None:
        await self._req("POST", f"/instances/{self.id_instance}/state-change",
                        {"from": from_, "to": to, "reason": reason})
